text/0000-router-helpers.md

```diff
@@ -0,0 +1,442 @@
+
+This RFC explodes out all of the functionality of `{{link-to}}` into a series of helpers. Since `{{link-to}}` was added to Ember it has grown naturally and unfortunately has added features that are non-performant and are not required by all applications. Because us this we plan to deprecate `{{link-to}}` and provide a [codemod](#migration-path) for migrating all instance of `{{link-to}}` to the corresponding helpers.
+
+### Deprecation of `{{query-params}}`
+
+This RFC introduces APIs that take `queryParams` as a named parameter and thus `{{query-params}}` is not needed as a stand alone helper.
+
+### Deprecation of Default Query Param Serialization
+
+As mentioned in the motivation, we want do not want to lookup the corresponding controller to serialize the default query param values. This means that any query params that you want to be serialized must be passed directly into the corresponding helper. As part of the migration **only** query params that were declared with `{{query-params}}` will be migrated.
+
+## Migration Path
+
+Since `{{link-to}}` is static we can write a codemod using [Ember Template Recast](https://github.com/ember-template-lint/ember-template-recast) to migrate the code. Below are numerous before and after examples of how the codemod would migrate.
+
+### Basic `{{link-to}}`
+
+**Before:**
+
+```hbs
+{{#link-to 'profile' class="profile"}}Profile{{/link-to}}
+{{link-to 'About' 'about'}}
+```
+
+**After:**
+
+```hbs
+<a href={{url-for 'profile'}} class="profile">Profile</a>
+<a href={{url-for 'about'}}>About</a>
+```
+
+### With Model `{{link-to}}`
+
+**Before:**
+
+```hbs
+{{#link-to 'profile' this.profile class="profile"}}Profile{{/link-to}}
+{{link-to 'About' 'about' this.contact}}
+```
+
+**After:**
+
+```hbs
+<a {{transition-to 'profile' this.profile}} class="profile">Profile</a>
+<a {{transition-to 'about' this.contact}}>About</a>
+```
+
+### With Query Params `{{link-to}}`
+
+**Before:**
+
+```hbs
+{{#link-to 'post' this.post (query-params order="CHRON")}}{{this.post.name}}{{/link-to}}
+```
+
+**After:**
+
+```hbs
+<a {{transition-to 'post' this.post queryParms=(hash order="CHRON")}}>{{this.post.name}}</a>
+```
+
+### With Replace `{{link-to}}`
+
+**Before:**
+
+```hbs
+{{#link-to 'post' this.post replace=true}}{{this.post.name}}{{/link-to}}
+```
+
+**After:**
+
+```hbs
+<a {{transition-to 'post' this.post replace=true}}>{{this.post.name}}</a>
+```
+
+One of the trickier parts about this migration is knowing how the autogenerated CSS classes are being used. Because of this, adding the route state helpers must explicitly be turned on in the codemod. For instance if you are making heavy use of the `.active` class, you will be suited best by turning pass the codemod the correct configuration to do a transform like the following:
+
+**Before**
+
+```hbs
+{{#link-to 'profile'}}Profile{{/link-to}}
+```
+
+**After**
+
+```hbs
+<a href={{url-for 'profile'}} class={{is-active 'active'}}>Profile</a>
```

Comment:
Is this correct for `{{is-active}}`? According to the signature you mentioned above, this should be `class={{if (is-active 'profile') 'active'}}` or is there some magic going on between `is-active` and `url-for` helper? I'd like the short syntax you used here, thought then it looks like overloading in that very special location?

Reply:
Agree here, I believe it should be updated to use the inline `if` snippet above.